src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -2670,5 +2670,225 @@ private interface INegator<T> where T : struct
+            where T : IComparable<T>
+        {
+            for (int i = length - 1; i >= 0; i--)
+            {
+                ref T current = ref Unsafe.Add(ref searchSpace, i);
+                if ((lowInclusive.CompareTo(current) <= 0) && (highInclusive.CompareTo(current) >= 0))
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        internal static int LastIndexOfAnyExceptInRange<T>(ref T searchSpace, T lowInclusive, T highInclusive, int length)
+            where T : IComparable<T>
+        {
+            for (int i = length - 1; i >= 0; i--)
+            {
+                ref T current = ref Unsafe.Add(ref searchSpace, i);
+                if ((lowInclusive.CompareTo(current) > 0) || (highInclusive.CompareTo(current) < 0))
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        internal static int LastIndexOfAnyInRangeUnsignedNumber<T>(ref T searchSpace, T lowInclusive, T highInclusive, int length)
+            where T : struct, IUnsignedNumber<T>, IComparisonOperators<T, T, bool> =>
+            LastIndexOfAnyInRangeUnsignedNumber<T, DontNegate<T>>(ref searchSpace, lowInclusive, highInclusive, length);
+
+        internal static int LastIndexOfAnyExceptInRangeUnsignedNumber<T>(ref T searchSpace, T lowInclusive, T highInclusive, int length)
+            where T : struct, IUnsignedNumber<T>, IComparisonOperators<T, T, bool> =>
+            LastIndexOfAnyInRangeUnsignedNumber<T, Negate<T>>(ref searchSpace, lowInclusive, highInclusive, length);
+
+        private static int LastIndexOfAnyInRangeUnsignedNumber<T, TNegator>(ref T searchSpace, T lowInclusive, T highInclusive, int length)
+            where T : struct, IUnsignedNumber<T>, IComparisonOperators<T, T, bool>
+            where TNegator : struct, INegator<T>
+        {
+            // T must be a type whose comparison operator semantics match that of Vector128/256.
+
+            if (!Vector128.IsHardwareAccelerated || length < Vector128<T>.Count)
+            {
+                T rangeInclusive = highInclusive - lowInclusive;
+                for (int i = length - 1; i >= 0; i--)
+                {
+                    ref T current = ref Unsafe.Add(ref searchSpace, i);
+                    if (TNegator.NegateIfNeeded((current - lowInclusive) <= rangeInclusive))
+                    {
+                        return i;
+                    }
+                }
+            }
+            else if (!Vector256.IsHardwareAccelerated || length < Vector256<T>.Count)
+            {
+                Vector128<T> inRangeVector, lowVector = Vector128.Create(lowInclusive), rangeVector = Vector128.Create(highInclusive - lowInclusive);
+                nint offset = length - Vector128<T>.Count;
+
+                // Loop until either we've finished all elements or there's a vector's-worth or less remaining.
+                while (offset > 0)
```

Comment:
Why not walk with the (managed) pointer through the searchSpace as before?
```c#
// Set current to last possible vector
ref T current = ref Unsafe.Add(ref searchSpace, (uint)(data.Length - Vector128<T>.Count));

do
{
    // ...

    current = ref Unsafe.Subtract(ref current, Vector128<T>.Count);
}
while (Unsafe.IsAddressGreaterThan(ref current, ref searchSpace));

// Remainder
Vector128<T> firstVec = Vector128.LoadUnsafe(ref searchSpace);
// ...
```

Perf-wise the addressing is backwards anyway, so cpu's prefetching won't be as effective as for forwards accesses.
With the pointer-walking it's guaranteed simple addressing mode though, but so far I never measured any difference for real workloads (except tiny (empty) loops) between simple adressing and the base + offset adressing.

Reply:
Becausr that has a GC hole.

Reply:
Ah, `current` could point before `searchSpace` and this isn't allowed for GC-tracked pointers. Thanks -- so I'll remember this.

---

Then the opposite side: why not use index / offset for the forward-scan too? For index of match computation the base would be already be by hand, no need for getting the byte-offset between searchSpace and current, then dividing by the size of T (though this will be executed once, so no perf-reason, but maybe less work for JIT, etc.).
As said, for addressing modes so far I could never measure any real difference between simple and base + offset.

Reply:
> current could point before searchSpace and this isn't allowed for GC-tracked pointers

To clarify, if Subtract were to move current to before the start of the object, and then between then and the IsAddressGreaterThan check the GC were to kick in and move the object, the searchSpace reference would be updated but current wouldn't be (since it's no longer pointing to the object).  That could then cause the comparison to return true, at which point we continue iterating using the erroneous value in current, and things go badly quickly.

> Then the opposite side: why not use index / offset for the forward-scan too?

I'm following suit with what the other IndexOf methods do here.  If we want to change them all to use the offset approach and it doesn't negatively impact perf, I'm fine with that.  If we want to change them all to instead pin and just use regular unmanaged pointers, I'm fine with that, too; that's less likely to lead to such GC holes and would likely result in cleaner code.  But I'm not going to consider either of those as part of this PR.

Reply:
Thanks for the clarification about GC holes.

Re addressing modes: makes sense.